Logbook-netty more extractions

sokomishalov opened the pull request:
According to Norman Maurer (netty maintainer), this is the only way to extract scheme:
https://stackoverflow.com/questions/36280376/get-base-uri-of-io-netty-handler-codec-http-httprequest#comment109701966_36281043

whiskeysierra commented:
Can you rebase onto main? I had to fix a CVE there which also breaks your build.

sokomishalov commented:
@whiskeysierra sure, done

whiskeysierra commented:
Ok, now the tests are failing due to coverage:

```
Warning:  Rule violated for class org.zalando.logbook.netty.Request: lines covered ratio is 0.9, but expected minimum is 1.0
Warning:  Rule violated for class org.zalando.logbook.netty.Request: branches covered ratio is 0.5, but expected minimum is 1.0
```

sokomishalov commented:
Added unit test to cover `org.zalando.logbook.netty.Request`
P.S. Fixed `CommonsLogFormatSinkTest` on hosts with another default host locale (caused wrong datetime formatting)
P.S.S. Replaced deprecated reactor-netty API in tests

whiskeysierra commented on logbook-netty/src/test/java/org/zalando/logbook/netty/LogbookClientHandlerTest.java:
```diff
@@ -48,11 +44,8 @@
-            .tcpConfiguration(tcpClient ->
-                    tcpClient.doOnConnected(connection ->
-                            connection.addHandlerLast(new LogbookClientHandler(logbook))))
+            .doOnConnected(connection -> connection.addHandlerLast(new LogbookClientHandler(logbook)))
```
The suppression was important, because the replacement method breaks backwards compatibility with prior versions of netty.

sokomishalov replied:
OK, I've reverted

whiskeysierra commented on logbook-netty/src/test/java/org/zalando/logbook/netty/LogbookClientHandlerTest.java:
```diff
@@ -33,11 +33,7 @@
-            .strategy(new TestStrategy())
-            .sink(new DefaultSink(new DefaultHttpLogFormatter(), writer))
-            .build();
+    private final Logbook logbook = Logbook.builder().strategy(new TestStrategy()).sink(new DefaultSink(new DefaultHttpLogFormatter(), writer)).build();
```
Can you revert the formatting?

sokomishalov replied:
sure

whiskeysierra commented on logbook-netty/src/test/java/org/zalando/logbook/netty/LogbookServerHandlerTest.java:
```diff
@@ -33,17 +33,10 @@
 final class LogbookServerHandlerTest {
 
     private final HttpLogWriter writer = mock(HttpLogWriter.class);
+    private final Logbook logbook = Logbook.builder().strategy(new TestStrategy()).sink(new DefaultSink(new DefaultHttpLogFormatter(), writer)).build();
```
Same here, reformatting is not needed, is it?

whiskeysierra commented on logbook-netty/src/test/java/org/zalando/logbook/netty/LogbookServerHandlerTest.java:
```diff
@@ -33,17 +33,10 @@
-            .tcpConfiguration(tcpServer ->
-                    tcpServer.doOnConnection(connection ->
-                            connection.addHandlerLast(new LogbookServerHandler(logbook))))
+            .doOnConnection(connection -> connection.addHandlerLast(new LogbookServerHandler(logbook)))
```
Suppression was intentional to preserve backwards compatibility.

sokomishalov replied:
reverted

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -114,6 +131,8 @@ void buffer(final ByteBuf content) {
         state.updateAndGet(state -> state.buffer(request, content));
     }
 
+
+
```
```suggestion
```

sokomishalov replied:
fixed

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -56,19 +64,29 @@ public String getMethod() {
     public String getHost() {
-        return request.headers().get(HOST, "unknown");
+        String host = request.headers().get(HOST);
+        if (host != null) {
```
I'd prefer to invert the expression/if.

sokomishalov replied:
done

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -56,19 +64,29 @@ public String getMethod() {
-        return request.headers().get(HOST, "unknown");
+        String host = request.headers().get(HOST);
+        if (host != null) {
+            int separator = host.indexOf(":");
+            return separator == -1 ? host : host.substring(0, separator);
```
Would be easier to understand if it were a small private method with a descriptive name, e.g. `stripPortIfNecessary`.

sokomishalov replied:
done

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -56,19 +64,29 @@ public String getMethod() {
+            int separator = host.indexOf(":");
+            return separator == -1 ? host : host.substring(0, separator);
+        } else {
+            return extractAddress().map(NetUtil::getHostname).orElse("unknown");
```
Is this safe to do? Does it to a DNS query? Would it actually report a hostname that the client didn't use?

sokomishalov replied:
Yes, it is
This method delegates to `InetSocketAddress::getHostString` on java >= 1.7
Replaced with explicit usage of this method.

whiskeysierra replied:
I'm still thinking, under which circumstances would this code execute and what does the hostname actually represent in that case? It's essentially a heuristic for the case that we didn't receive a `Host` header, right? We then try to do a reverse lookup of the channel's ip address to find the most likely hostname?

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -56,19 +64,29 @@ public String getMethod() {
+    }
+
+    private Optional<InetSocketAddress> extractAddress() {
+        SocketAddress address = origin == Origin.LOCAL ? context.channel().remoteAddress() : context.channel().localAddress();
```
```suggestion
        final Channel channel = context.channel();
        final SocketAddress address = origin == LOCAL ? channel.remoteAddress() : channel.localAddress();
```

whiskeysierra replied:
Static import for Origin allows to use LOCAL directly.

sokomishalov replied:
fixed

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -56,19 +64,29 @@ public String getMethod() {
-        return request.headers().get(HOST, "unknown");
+        String host = request.headers().get(HOST);
+        if (host != null) {
+            int separator = host.indexOf(":");
```
```suggestion
            final int separator = host.indexOf(":");
```

sokomishalov replied:
fixed

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -56,19 +64,29 @@ public String getMethod() {
     @Override
     public String getHost() {
-        return request.headers().get(HOST, "unknown");
+        String host = request.headers().get(HOST);
```
```suggestion
        final String host = request.headers().get(HOST);
```

sokomishalov replied:
fixed

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -56,19 +64,29 @@ public String getMethod() {
     public String getScheme() {
-        // TODO pick the real one
-        return "http";
+        SslHandler handler = context.channel().pipeline().get(SslHandler.class);
```
```suggestion
        final SslHandler handler = context.channel().pipeline().get(SslHandler.class);
```

sokomishalov replied:
fixed

whiskeysierra commented on logbook-netty/src/main/java/org/zalando/logbook/netty/Request.java:
```diff
@@ -89,13 +107,12 @@ public HttpHeaders getHeaders() {
     @Override
     public String getContentType() {
-        return request.headers().get(CONTENT_TYPE);
+        return Optional.ofNullable(HttpUtil.getMimeType(request)).map(CharSequence::toString).orElse(null);
```
```suggestion
        return Objects.toString(HttpUtil.getMimeType(request), null);
```

whiskeysierra commented on logbook-core/src/test/java/org/zalando/logbook/CommonsLogFormatSinkTest.java:
```diff
@@ -26,6 +28,11 @@
 
+    @BeforeEach
+    void setUp() {
+        Locale.setDefault(Locale.ENGLISH);
```
Do you happen to have the error (message) at hand that you experienced without this?

sokomishalov replied:
![image](https://user-images.githubusercontent.com/21218309/114193010-c833b080-9956-11eb-8fd6-9895fe7593c8.png)
As you can see, month letters are cyrillic on my machine with the default locale (Russian in my case) and it causes test failures.
And I decided that changing formatter locale in `CommonsLogFormatSink` would be breaking change so I added this setup.

whiskeysierra replied:
Actually, I believe you discovered a bug. CLF is using US date time format, afaik.

sokomishalov replied:
Well, I supposed the same way as you. But specification says [here](https://en.wikipedia.org/wiki/Common_Log_Format) and [here](https://strftime.org/):
```
[10/Oct/2000:13:55:36 -0700] is the date, time, and time zone that the request was received, by default in strftime format %d/%b/%Y:%H:%M:%S %z.
.....
%b - Month as **locale’s** abbreviated name (i.e. Sep)
```

whiskeysierra commented on logbook-netty/src/test/java/org/zalando/logbook/netty/LogbookClientHandlerTest.java:
```diff
@@ -39,6 +39,7 @@
             .sink(new DefaultSink(new DefaultHttpLogFormatter(), writer))
             .build();
 
+
```
```suggestion
```

whiskeysierra commented:
👍

whiskeysierra commented:
@sokomishalov Thank you, for your contributions! It's highly appreciated. Sorry that it took so long and thanks for engaging in the review! 📣 

I just released your changes as 2.7.0 to maven central.
